### PR TITLE
 Follow spring rule, Only rollback on setting exceptions

### LINKED ISSUE #1097: ConnectionProxy processLocalCommitWithGlobalLocks fail when commit secondly
- [ ] I have searched the [issues](https://github.com/seata/seata/issues) of this repository and believe that this is not a duplicate.

### Ⅰ. Issue Description
checkLock(context.buildLockKeys());  fail.

### Ⅱ. Describe what happened

AbstractDMLBaseExecutor.executeAutoCommitTrue will call connectionProxy.commit() and clear lockKey(context.reset()); finally connectionProxy.setAutoCommit(true) will be called.

However,  change autocommit from false to true will call commit twice, lockKey is empty when checkLock.

    `public void setAutoCommit(boolean autoCommit) throws SQLException {
        if ((autoCommit) && !getAutoCommit()) {
            // change autocommit from false to true, we should commit() first according to JDBC spec.
            commit();
        }
        targetConnection.setAutoCommit(autoCommit);
    }`

### Ⅳ. How to reproduce it (as minimally and precisely as possible)

1. xxx
2. xxx
3. xxx

### Ⅵ. Environment:

- JDK version :
- OS :
- Others:

### LINKED ISSUE #1098: BaseTransactionalExecutor.prepareUndoLog should consider reset ConnectionContext when retry
- [ ] I have searched the [issues](https://github.com/seata/seata/issues) of this repository and believe that this is not a duplicate.

### Ⅰ. Issue Description
lockKeysBuffer & sqlUndoItemsBuffer contain dup key and item when prepareUndoLog call multi time.

### Ⅱ. Describe what happened

AbstractDMLBaseExecutor.executeAutoCommitTrue  line: 90

            while (true) {
                try {
                    result = executeAutoCommitFalse(args);
                    connectionProxy.commit();
                    break;
                } catch (LockConflictException lockConflict) {
                    connectionProxy.getTargetConnection().rollback();
                    lockRetryController.sleep(lockConflict);
                }
            }

when seata transaction lock, executeAutoCommitTrue will retry  executeAutoCommitFalse() 30 time when LockConflictException happen. This means lockKeysBuffer & sqlUndoItemsBuffer will dup.

> [20:01:09:083]datasource.ConnectionProxy:94[netty-server-handler-0] DEBUG GlobalLocks check LockKeys=storage:1,2,3,4;storage:1,2,3,4;storage:1,2,3,4;storage:1,2,3,4;storage:1,2,3,4;storage:1,2,3,4;storage:1,2,3,4;storage:1,2,3,4;storage:1,2,3,4;storage:1,2,3,4;storage:1,2,3,4;storage:1,2,3,4;storage:1,2,3,4;storage:1,2,3,4;storage:1,2,3,4;storage:1,2,3,4;storage:1,2,3,4;storage:1,2,3,4;storage:1,2,3,4;storage:1,2,3,4;storage:1,2,3,4;storage:1,2,3,4;storage:1,2,3,4;storage:1,2,3,4;storage:1,2,3,4;storage:1,2,3,4;storage:1,2,3,4;storage:1,2,3,4;storage:1,2,3,4;storage:1,2,3,4;storage:1,2,3,4

### Ⅳ. How to reproduce it (as minimally and precisely as possible)

1. xxx
2. xxx
3. xxx

### Ⅵ. Environment:

- JDK version :
- OS :
- Others:

### LINKED ISSUE #1156: GlobalTransactional rollbackFor will rollback on unset exception
- [ ] I have searched the [issues](https://github.com/seata/seata/issues) of this repository and believe that this is not a duplicate.

### Ⅰ. Issue Description

Set: rollbackFor = {RuntimeException.class}, if throw Exception then Transaction will be rollback. It's not same with spring transaction: only throw RuntimeException will rollback.

### Ⅱ. Describe what happened

  If there is an exception, please attach the exception trace:

```
Just paste your stack trace here!
```

### Ⅳ. How to reproduce it (as minimally and precisely as possible)

1. xxx
2. xxx
3. xxx

### Ⅵ. Environment:

- JDK version :
- OS :
- Others:

### PR DESCRIPTION
**Pull request by zhenjing**
### Ⅰ. Describe what this PR did

Follow spring rule, Only rollback on setting exceptions

### Ⅱ. Does this pull request fix one issue?

 
Fixed #1156 

Fixed #1098

Fixed #1097

### CODE REVIEW

File: rm-datasource/src/main/java/io/seata/rm/datasource/ConnectionProxy.java [hunk 1]
```diff
@@ -25,6 +25,7 @@
 import io.seata.rm.datasource.exec.LockConflictException;
 import io.seata.rm.datasource.undo.SQLUndoLog;
 import io.seata.rm.datasource.undo.UndoLogManager;
+import org.apache.commons.lang.StringUtils;
```
Comment (leizhiyuan) [inline rm-datasource/src/main/java/io/seata/rm/datasource/ConnectionProxy.java]: use io.seata.common.util.StringUtils

File: tm/src/main/java/io/seata/tm/api/transaction/TransactionInfo.java [hunk 2]
```diff
@@ -60,8 +60,16 @@ public boolean rollbackOn(Throwable ex) {
...
         RollbackRule winner = null;
         int deepest = Integer.MAX_VALUE;
+        boolean setRollback = false;
 
         if (this.rollbackRules != null) {
+            for (RollbackRule rule : this.rollbackRules){
+                if(rule instanceof NoRollbackRule){
+                    continue;
+                }
+                setRollback = true;
```
Comment (leizhiyuan) [inline tm/src/main/java/io/seata/tm/api/transaction/TransactionInfo.java]: execute many times?

File: rm-datasource/src/main/java/io/seata/rm/datasource/ConnectionProxy.java (hunk 1 above)
Comment (zhenjing) [reply rm-datasource/src/main/java/io/seata/rm/datasource/ConnectionProxy.java]: done

File: tm/src/main/java/io/seata/tm/api/transaction/TransactionInfo.java (hunk 2 above)
Comment (zhenjing) [reply tm/src/main/java/io/seata/tm/api/transaction/TransactionInfo.java]: done

Comment (github-ygy) [reply tm/src/main/java/io/seata/tm/api/transaction/TransactionInfo.java]: What you have modified and what you originally expressed, I can’t see any changes, the code is more complicated.

File: tm/src/main/java/io/seata/tm/api/transaction/TransactionInfo.java [hunk 3]
```diff
@@ -68,9 +69,15 @@ public boolean rollbackOn(Throwable ex) {
...
             }
         }
 
-        return winner == null || !(winner instanceof NoRollbackRule);
+        return (!setRollback && winner == null)
+            || (winner != null && !(winner instanceof NoRollbackRule));
```
Comment (github-ygy) [inline tm/src/main/java/io/seata/tm/api/transaction/TransactionInfo.java]: i can't see any changes

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/AbstractDMLBaseExecutor.java [hunk 4]
```diff
@@ -88,9 +87,10 @@ protected T executeAutoCommitTrue(Object[] args) throws Throwable {
         LockRetryController lockRetryController = new LockRetryController();
         try {
             connectionProxy.setAutoCommit(false);
+
+            result = executeAutoCommitFalse(args);
```
Comment (slievrly) [inline rm-datasource/src/main/java/io/seata/rm/datasource/exec/AbstractDMLBaseExecutor.java]: During the retry, the before-image data may change. This is wrong.

File: rm-datasource/src/main/java/io/seata/rm/datasource/ConnectionProxy.java [hunk 5]
```diff
@@ -163,7 +164,11 @@ public void commit() throws SQLException {
 
     private void processLocalCommitWithGlobalLocks() throws SQLException {
 
-        checkLock(context.buildLockKeys());
+        String lockKeys = getContext().buildLockKeys();
```
Comment (slievrly) [inline rm-datasource/src/main/java/io/seata/rm/datasource/ConnectionProxy.java]: Separate this modification into a separate pr

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/AbstractDMLBaseExecutor.java (hunk 4 above)
Comment (zhenjing) [reply rm-datasource/src/main/java/io/seata/rm/datasource/exec/AbstractDMLBaseExecutor.java]: 1. prepareUndoLog() should clear undolog in this case.
2. before-image cost much time in batch-situation. 

I will remove this changed line, discuss it in other issue.

Comment (slievrly) [reply rm-datasource/src/main/java/io/seata/rm/datasource/exec/AbstractDMLBaseExecutor.java]: 1. yeah, undolog should add in Set but List.
2. batch solution should use autocommit=false
